Marlin/src/lcd/extui/lib/mks_ui/draw_touch_calibration.cpp

```diff
@@ -0,0 +1,120 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2020 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+#include "../../../../inc/MarlinConfigPre.h"
+
+#if BOTH(HAS_TFT_LVGL_UI, TOUCH_SCREEN_CALIBRATION)
+
+#include "draw_ui.h"
+#include "draw_touch_calibration.h"
+#include <lv_conf.h>
+
+#include "../../../../inc/MarlinConfig.h"
+#include "../../../tft_io/touch_calibration.h"
+#include "SPI_TFT.h"
+
+static lv_obj_t *scr;
+static lv_obj_t *status_label;
+
+static void event_handler(lv_obj_t *obj, lv_event_t event);
+
+enum {
+  ID_TC_RETURN = 1
+};
+
+static void drawCross(uint16_t x, uint16_t y, uint16_t color) {
+  SPI_TFT.tftio.set_window(x - 15, y, x + 15, y);
+  SPI_TFT.tftio.WriteMultiple(color, 31);
+  SPI_TFT.tftio.set_window(x, y - 15, x, y + 15);
+  SPI_TFT.tftio.WriteMultiple(color, 31);
+}
+
+void lv_update_touch_calibration_screen() {
+  uint16_t x, y;
+
+  calibrationState calibration_stage = touch_calibration.get_calibration_state();
+  if (calibration_stage == CALIBRATION_NONE) {
+    // start and clear screen
+    calibration_stage = touch_calibration.calibration_start();
+  }
+  else {
+    // clear last cross
+    x = touch_calibration.calibration_points[_MIN(calibration_stage - 1, CALIBRATION_BOTTOM_RIGHT)].x;
+    y = touch_calibration.calibration_points[_MIN(calibration_stage - 1, CALIBRATION_BOTTOM_RIGHT)].y;
+    drawCross(x, y, LV_COLOR_BACKGROUND.full);
+  }
+
+  const char *str = nullptr;
+  if (calibration_stage < CALIBRATION_SUCCESS) {
+    // handle current state
+    switch (calibration_stage) {
+      case CALIBRATION_TOP_LEFT: str = "Top Left"; break;
+      case CALIBRATION_BOTTOM_LEFT: str = "Bottom Left"; break;
+      case CALIBRATION_TOP_RIGHT:  str = "Top Right"; break;
+      case CALIBRATION_BOTTOM_RIGHT: str = "Bottom Right"; break;
+      default: break;
+    }
+
+    x = touch_calibration.calibration_points[calibration_stage].x;
+    y = touch_calibration.calibration_points[calibration_stage].y;
+    drawCross(x, y, LV_COLOR_WHITE.full);
+  }
+  else {
+    // end calibration
+    str = calibration_stage == CALIBRATION_SUCCESS ? "Calibration Completed" : "Calibration Failed, send M995 to try again";
```

Comment:
For these screens we use `GET_TEXT_F` to get strings for the current language. It looks like this and some other ExtUI screens are still using fixed English strings where localized strings would be better.

Reply:
well.... the origin of this is: to create a generic touch calibration screen, we would have our separated font to draw the screen, and that would be limited to save flash..... but, in the middle of the dev, I changed it, and made the UI responsible to draw itself. This now allow us to use i18n here. I will update the code.